ngrok/src/internals/raw_session.rs

```diff
@@ -290,15 +334,72 @@ impl RpcClient {
     }
 }
 
+pub const NOT_IMPLEMENTED: &str = "the agent has not defined a callback for this operation";
+
+async fn handle_req<T>(
+    handler: Option<Arc<dyn CommandHandler<T>>>,
+    mut stream: TypedStream,
+) -> Result<(), Either<io::Error, serde_json::Error>>
+where
+    T: DeserializeOwned + Debug + 'static,
+{
+    let res = async {
+        debug!("reading request from stream");
+        let mut buf = vec![];
+        let req = serde_json::from_value(loop {
+            let mut tmp = vec![0u8; 256];
+            let bytes = stream.read(&mut tmp).await.map_err(Either::Left)?;
+            buf.extend_from_slice(&tmp[..bytes]);
+
+            if let Ok(obj) = serde_json::from_slice::<serde_json::Value>(&buf) {
+                break obj;
+            }
+        })
```

Comment:
This does feel odd. But requests will almost always fit in 256 bytes so likely won't be doing much extra work really?

Reply:
Yeah, common case would still be to pull all of that message in a single non-blocking read request. Json parse would fail until the whole message is there, so degenerate case is a byte-by-byte loop and parse but in practice it'll likely be a one-shot. Would be nice to do piecemeal parse rather than full, but an optimization unlikely to be hit, on a function infrequently called, would guess there's bigger fish to fry.